### PR TITLE
Making `DataProvider: Sync + Send`

### LINKED ISSUE #1852: Make DataPayload be Send/Sync on supported platforms
Currently `DataPayload` has an `Rc` inside of it, which makes it not support Send/Sync.

There are various ways to solve this problem, none of which are perfect:

1. Make the Cart backend configurable via the signature of DataPayload
    - I would really like to avoid adding more type parameters to DataPayload; it goes against the train of how Yoke is supposed to work
2. Replace the Rc with a Box
    - Then we have to allocate a lot more often, so this is not a good idea
3. Use a separate type hierarchy for send/sync versus single-threaded
    - This would bubble down to everything that uses DataPayload and would further increase the complexity of the data provider
4. Make `Rc` vs. `Arc` build-time configurable via a feature on the icu_provider crate
    - It's generally considered bad practice to have features that are not just additive

I lean toward (4). More specifically:

- We add an `"arc"` feature to `icu_provider`
- This feature changes the `Rc` to an `Arc`
- DataPayload signatures that take an `Rc` continue to take an `Rc`, and convert it to an `Arc` internally
- Additional methods that take an `Arc` directly can be enabled when the `"arc"` feature is enabled

Needs approval from:

- [ ] @robertbastian 
- [ ] @Manishearth

### PR DESCRIPTION
**Pull request by robertbastian**
Fixes #1852

### CODE REVIEW

File: provider/core/src/data_provider.rs
Comment (sffc) [inline provider/core/src/data_provider.rs:20]: Question: Are you able to put `type Foo = Bar` in a `cfg` or do you need to use the macro?

File: provider/core/src/data_provider.rs
Comment (sffc) [inline provider/core/src/data_provider.rs:188]: Issue, here and elsewhere: Enabling/disabling features should not change the signature of any functions. If there's code that is built to rely on a feature being turned off, and another crate turns it on, you break things.

In other words, the change should be internal only, adding the `Send + Sync` to DataPayload but not changing any visible function signatures.

Not sure of the right solution right now, but thought I'd flag it first so we can discuss

**Review by sffc — CHANGES_REQUESTED** [review]

File: provider/core/src/data_provider.rs
Comment (robertbastian) [reply provider/core/src/data_provider.rs:188]: That's why I changed the signature to `B: Into<rc_type!()>`. `Rc` and `Arc` have the same `From` implementations.

Comment (sffc) [reply provider/core/src/data_provider.rs:188]: Issues with this:

1. We cannot assume that types that implement `Into<Rc<[u8]>>` necessarily implement `Into<Arc<[u8]>>` and vice-versa
2. Replying on `Into` is a footgun because it could cause the data to be cloned, and we really want to avoid this function having to clone any data

I suggested adding a `RcWrap` type, or basically an ICU4X-specific Rc that clients should use.

File: provider/blob/Cargo.toml
Comment (sffc) [inline provider/blob/Cargo.toml:51]: Nit: You don't need a `"sync"` feature on this crate. You only need it on `icu_provider`.

File: provider/core/Cargo.toml
Comment (sffc) [inline provider/core/Cargo.toml:38]: Discussion: I think `"sync"` should not be default.

File: provider/core/src/data_provider.rs
Comment (sffc) [inline provider/core/src/data_provider.rs:173]: Nit: Always include safety docs. Each trait should have a section that tells you what invariants need to be upheld.

File: provider/core/src/data_provider.rs
Comment (sffc) [inline provider/core/src/data_provider.rs:245]: Nit: You currently are inconsistent: you have `try_from_rc_buffer` and `try_from_buffer_badly`.

I have a preference for keeping the `_rc` in the function name.

File: provider/core/src/data_provider.rs
Comment (sffc) [inline provider/core/src/data_provider.rs:249]: Nit: I think you should not need to call `.into()` here

File: provider/core/src/lib.rs
Comment (sffc) [inline provider/core/src/lib.rs:171]: Suggestion: If we add this to the prelude, it should have a more unique name. Alternatively, just move it to the top level.

**Review by sffc — CHANGES_REQUESTED** [review]

File: provider/blob/Cargo.toml
Comment (robertbastian) [reply provider/blob/Cargo.toml:51]: I think it's weird not to have it here. `BlobDataProvider` is the concrete type that clients will be using, and it's weird if that changes its `Sync`ness depending on the features of another crate.

File: provider/core/Cargo.toml
Comment (robertbastian) [reply provider/core/Cargo.toml]: agree

File: provider/blob/Cargo.toml
Comment (Manishearth) [reply provider/blob/Cargo.toml:51]: I think `BlobDataProvider` is not the _only_ crate like this, and it's reasonably normal for dependency crate features to be toggled this way. The type comes from `icu_provider`, people will be importing `icu_provider` (it's not a secret internal crate), they can enable the feature there directly.

The feature should go on BlobDataProvider if BlobDataProvider also does something special for it.

File: provider/core/src/dynutil.rs
Comment (sffc) [inline provider/core/src/dynutil.rs:255]: Nit: `std` should be `alloc` (even, IMO, if it is in a context that requires std)

**Review by sffc — APPROVED** [review]
I think there are some docs links errors

**Comment by sffc** [comment]
Also, please wait for an explicit approval from @Manishearth

File: provider/core/src/dynutil.rs
Comment (robertbastian) [reply provider/core/src/dynutil.rs:255]: `alloc` requires `extern crate alloc`. This is only used by datagen which is `std`

**Comment by robertbastian** [comment]
@Manishearth ptal, I need this for datagen improvements.

**Review by Manishearth — APPROVED** [review]

**Pull request merged**
